### PR TITLE
[experimental]planner: change semi join to inner join when decorrelaing

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve? 

open more probability for optimization.
A experimental feature.

### What is changed and how it works?

If the join is non-corrleated. Try to rewrite to inner join if it's a simple semi join.

### Check List 

Tests 

 - Unit test

Related changes
Currently none.

### CODE REVIEW

File: planner/core/logical_plan_test.go [hunk 1]
```diff
@@ -159,23 +159,23 @@ func (s *testPlanSuite) TestPredicatePushDown(c *C) {
 		},
 		{
 			sql:  "select a from t where exists(select 1 from t as x where x.a = t.a) and exists(select 1 from t as x where x.a = t.a)",
-			best: "Join{Join{DataScan(t)->DataScan(x)}(test.t.a,test.x.a)->DataScan(x)}(test.t.a,test.x.a)->Projection",
+			best: "Join{Join{DataScan(t)->DataScan(x)->Projection}(test.t.a,test.x.a)->Projection->DataScan(x)->Projection}(test.t.a,test.x.a)->Projection->Projection",
 		},
 		{
 			sql:  "select * from (select a, b, sum(c) as s from t group by a, b) k where k.a > k.b * 2 + 1",
-			best: "DataScan(t)->Aggr(sum(test.t.c),firstrow(test.t.a),firstrow(test.t.b))->Projection->Projection",
```
Comment (qw4990) [inline planner/core/logical_plan_test.go]: Why this `Aggregation` is eliminated?

File: planner/core/optimizer.go [hunk 2]
```diff
@@ -66,6 +68,7 @@ var optRuleList = []logicalOptRule{
 // logicalOptRule means a logical optimizing rule, which contains decorrelate, ppd, column pruning, etc.
 type logicalOptRule interface {
 	optimize(LogicalPlan) (LogicalPlan, error)
+	getFlag() uint64
```
Comment (qw4990) [inline planner/core/optimizer.go]: Please add some comments...

**Comment by eurekaka** [comment]
Could you please shed some light on the status of this attempt?

**Comment by zz-jason** [comment]
@winoros please update to the master branch and resolve conflicts.

**Comment by zz-jason** [comment]
I'm going to close this PR due to outdated, fell free to reopen it if you are going to process it again in the future.

**Pull request closed without merging**